File: partition.java
// Time Complexity : O(n)
// Space Complexity :O(n)
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
class Solution {
    public List<Integer> partitionLabels(String S) {
        HashMap<Character, Integer> map = new HashMap<>();
        
        for(int i = 0; i< S.length();i++)
            map.put(S.charAt(i),i);

        int start = 0, end = 0;
        
        ArrayList<Integer> output = new ArrayList <>();
        
        for(int i = 0; i < S.length(); i++){
            end = Math.max(end, map.get(S.charAt(i)));
            
            if(end == i){
                output.add(end - start + 1);
                start = end + 1;
                end = start;
            }
        }
            
        return output;
    }
}

File: partition.swift
// Time Complexity : 0(n)
// Space Complexity : 0(n)
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach

    func partitionLabels(_ s: String) -> [Int] {
       if s == nil || s.count == 0 {
           return []
       } 
        var hashmap = [Character:Int]()
        var output = [Int]()
        var sArray = Array(s)
        for i in 0..<s.count {
            hashmap[sArray[i]] = i
        }
        var start = 0
        var end = hashmap[sArray[0]]! 
        for i in 0..<s.count {
             if hashmap[sArray[i]]! > end  {
                 end = hashmap[sArray[i]]!
             }
             if end == i {
                 output.append(end - start + 1)
                 start = end + 1
             }
        }
        return output
    }
